The job **`x86_64-gnu-gcc`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/21565859764/job/62137134243) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/62137134243) [(plain)](https://github.com/rust-lang/rust/commit/b3be537510210d388e423b84e974777c78763016/checks/62137134243/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
REPOSITORY                                   TAG       IMAGE ID       CREATED      SIZE
ghcr.io/dependabot/dependabot-updater-core   latest    bcec0b4e062b   6 days ago   783MB
=> Removing docker images...
Deleted Images:
untagged: ghcr.io/dependabot/dependabot-updater-core:latest
untagged: ghcr.io/dependabot/dependabot-updater-core@sha256:b662be51f7b8ef7e2c8464428f14e49cb79c36aa9afb7ecb9221dfe0f507050c
deleted: sha256:bcec0b4e062b5ffe11cc1c2729558c0cd96621c0271ab5e97ff3a56e0c25045a
deleted: sha256:64e147d5e54d9be8b8aa322e511cda02296eda4b8b8d063c6a314833aca50e29
deleted: sha256:5cba409bb463f4e7fa1a19f695450170422582c1bc7c0e934d893b4e5f558bc6
deleted: sha256:cddc6ebd344b0111eaab170ead1dfda24acdfe865ed8a12599a34d338fa8e28b
deleted: sha256:2412c3f334d79134573cd45e657fb6cc0abd75bef3881458b0d498d936545c8d
---
test [ui] tests/ui/asm/aarch64/type-check-2-2.rs ... ignored, only executed when the architecture is aarch64
test [ui] tests/ui/asm/aarch64/type-check-2.rs ... ignored, only executed when the architecture is aarch64
test [ui] tests/ui/asm/aarch64/type-check-3.rs ... ignored, only executed when the architecture is aarch64
test [ui] tests/ui/asm/aarch64/type-f16.rs ... ignored, only executed when the architecture is aarch64
test [ui] tests/ui/asm/aarch64v8r.rs#hf ... ignored, gcc backend is marked as ignore
test [ui] tests/ui/asm/aarch64v8r.rs#sf ... ignored, gcc backend is marked as ignore
test [ui] tests/ui/asm/arm-low-dreg.rs ... ignored, gcc backend is marked as ignore
test [ui] tests/ui/array-slice-vec/vec-tail-matching.rs ... ok
test [ui] tests/ui/asm/bad-template.rs#aarch64 ... ignored, gcc backend is marked as ignore
test [ui] tests/ui/asm/bad-template.rs#x86_64 ... ignored, gcc backend is marked as ignore
test [ui] tests/ui/array-slice-vec/vector-no-ann-2.rs ... ok
---
test [ui] tests/ui/extern/issue-36122-accessing-externed-dst.rs ... ok
test [ui] tests/ui/extern/issue-95829.rs ... ok
test [ui] tests/ui/extern/not-in-block.rs ... ok
test [ui] tests/ui/extern/unsized-extern-derefmove.rs ... ok
test [ui] tests/ui/extern/lgamma-linkage.rs ... ok
test [ui] tests/ui/feature-gates/allow-features-empty.rs ... ok
test [ui] tests/ui/extern/issue-80074.rs ... ok
test [ui] tests/ui/feature-gates/allow-features.rs ... ok
test [ui] tests/ui/extern/windows-tcb-trash-13259.rs ... ok
test [ui] tests/ui/feature-gates/bench.rs ... ok
---
test [ui] tests/ui/imports/ambiguous-2.rs ... ok
test [ui] tests/ui/imports/ambiguous-7.rs ... ok
test [ui] tests/ui/imports/ambiguous-4.rs ... ok
test [ui] tests/ui/imports/ambiguous-9.rs ... ok
test [ui] tests/ui/imports/ambiguous-panic-glob-vs-multiouter.rs ... ok
test [ui] tests/ui/imports/ambiguous-panic-globvsglob.rs ... ok
test [ui] tests/ui/imports/ambiguous-panic-no-implicit-prelude.rs ... ok
test [ui] tests/ui/imports/ambiguous-8.rs ... ok
test [ui] tests/ui/imports/ambiguous-glob-vs-expanded-extern.rs ... ok
test [ui] tests/ui/imports/ambiguous-panic-non-prelude-core-glob.rs ... ok
test [ui] tests/ui/imports/ambiguous-panic-non-prelude-std-glob.rs ... ok
---
---- [ui] tests/ui/asm/x86_64/global_asm_escape.rs stdout ----

error: test compilation failed although it shouldn't!
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/asm/x86_64/global_asm_escape.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "-Zcodegen-backend=gcc" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "-O" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/asm/x86_64/global_asm_escape/a" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers"
stdout: none
--- stderr -------------------------------
libgccjit.so: error: : gcc_jit_context_new_function: name "exit@GLIBC_2.2.5" contains invalid character: '@'
gcc_jit_function_add_attribute: NULL func
error: linking with `cc` failed: exit status: 1
   |
   = note:  "cc" "-m64" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/asm/x86_64/global_asm_escape/rustcWWbcbw/symbols.o" "<1 object files omitted>" "-Wl,--as-needed" "-Wl,-Bdynamic" "<sysroot>/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-4b847fc65409bf4c.so" "-Wl,-Bstatic" "<sysroot>/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-*.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/asm/x86_64/global_asm_escape/rustcWWbcbw/raw-dylibs" "-B<sysroot>/lib/rustlib/x86_64-unknown-linux-gnu/bin/gcc-ld" "-B<sysroot>/lib/rustlib/x86_64-unknown-linux-gnu/bin/gcc-ld" "-fuse-ld=lld" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "<sysroot>/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/asm/x86_64/global_asm_escape/a" "-Wl,--gc-sections" "-pie" "-Wl,-z,relro,-z,now" "-Wl,-O1" "-Wl,--strip-debug" "-nodefaultlibs" "-Wl,-rpath,$ORIGIN/../../../../../stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib,--enable-new-dtags,-z,origin"
   = note: some arguments are omitted. use `--verbose` to show all linker arguments
   = note: rust-lld: error: cannot open /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/asm/x86_64/global_asm_escape/a.global_asm_escape.48ee2b60b27cace7-cgu.0.rcgu.o: No such file or directory
           collect2: error: ld returned 1 exit status
           

error: aborting due to 1 previous error
------------------------------------------

---- [ui] tests/ui/asm/x86_64/global_asm_escape.rs stdout end ----

failures:
    [ui] tests/ui/asm/x86_64/global_asm_escape.rs

test result: FAILED. 19669 passed; 1 failed; 839 ignored; 0 measured; 0 filtered out; finished in 792.86s
```

</details>

For more information how to resolve CI failures of this job, visit this [link](https://rustc-dev-guide.rust-lang.org/tests/codegen-backend-tests/cg_gcc.html).

--- a/compiler/rustc_codegen_llvm/src/asm.rs
+++ b/compiler/rustc_codegen_llvm/src/asm.rs
@@ -399,7 +399,7 @@ impl<'tcx> AsmCodegenMethods<'tcx> for CodegenCx<'_, 'tcx> {
         for piece in template {
             match *piece {
                 InlineAsmTemplatePiece::String(ref s) => template_str.push_str(s),
-                InlineAsmTemplatePiece::Placeholder { operand_idx, modifier: _, span: _ } => {
+                InlineAsmTemplatePiece::Placeholder { operand_idx, modifier: _, span } => {
                     match operands[operand_idx] {
                         GlobalAsmOperandRef::Const { ref string } => {
                             // Const operands get injected directly into the
@@ -414,7 +414,7 @@ impl<'tcx> AsmCodegenMethods<'tcx> for CodegenCx<'_, 'tcx> {
                                 llvm::LLVMRustGetMangledName(llval, s);
                             })
                             .expect("symbol is not valid UTF-8");
-                            template_str.push_str(&symbol);
+                            template_str.push_str(&escape_symbol_name(self, symbol, span));
                         }
                         GlobalAsmOperandRef::SymStatic { def_id } => {
                             let llval = self
@@ -428,7 +428,7 @@ impl<'tcx> AsmCodegenMethods<'tcx> for CodegenCx<'_, 'tcx> {
                                 llvm::LLVMRustGetMangledName(llval, s);
                             })
                             .expect("symbol is not valid UTF-8");
-                            template_str.push_str(&symbol);
+                            template_str.push_str(&escape_symbol_name(self, symbol, span));
                         }
                     }
                 }
@@ -1389,4 +1389,43 @@ fn llvm_fixup_output_type<'ll, 'tcx>(
         ) if s.primitive() == Primitive::Float(Float::F64) => cx.type_vector(cx.type_f64(), 2),
         _ => layout.llvm_type(cx),
     }
+}
+
+fn escape_symbol_name<'ll, 'tcx>(cx: &CodegenCx<'ll, 'tcx>, symbol: String, span: Span) -> String {
+    use rustc_target::spec::{Arch, BinaryFormat};
+    if !symbol.is_empty()
+        && symbol.chars().all(|c| matches!(c, '0'..='9' | 'A'..='Z' | 'a'..='z' | '_' | '$' | '.'))
+    {
+        return symbol;
+    }
+    if cx.tcx.sess.target.binary_format == BinaryFormat::Xcoff {
+        cx.tcx.sess.dcx().span_fatal(
+            span,
+            format!(
+                "symbol escaping is not supported for the binary format {}",
+                cx.tcx.sess.target.binary_format
+            ),
+        );
+    }
+    if cx.tcx.sess.target.arch == Arch::Nvptx64 {
+        cx.tcx.sess.dcx().span_fatal(
+            span,
+            format!(
+                "symbol escaping is not supported for the architecture {}",
+                cx.tcx.sess.target.arch
+            ),
+        );
+    }
+    let mut escaped_symbol = String::new();
+    escaped_symbol.push('\"');
+    for c in symbol.chars() {
+        match c {
+            '\n' => escaped_symbol.push_str("\\\n"),
+            '"' => escaped_symbol.push_str("\\\""),
+            '\\' => escaped_symbol.push_str("\\\\"),
+            c => escaped_symbol.push(c),
+        }
+    }
+    escaped_symbol.push('\"');
+    escaped_symbol
 }

--- a/tests/ui/asm/x86_64/global_asm_escape.rs
+++ b/tests/ui/asm/x86_64/global_asm_escape.rs
@@ -1,0 +1,17 @@
+//@ run-pass
+//@ only-x86_64-unknown-linux-gnu
+//@ ignore-backends: gcc
+
+// https://github.com/rust-lang/rust/issues/151950
+
+unsafe extern "C" {
+    #[link_name = "exit@GLIBC_2.2.5"]
+    safe fn exit(status: i32) -> !;
+    safe fn my_exit(status: i32) -> !;
+}
+
+core::arch::global_asm!(".global my_exit", "my_exit:", "jmp {}", sym exit);
+
+fn main() {
+    my_exit(0);
+}

--- a/tests/ui/asm/x86_64/naked_asm_escape.rs
+++ b/tests/ui/asm/x86_64/naked_asm_escape.rs
@@ -1,0 +1,32 @@
+//@ build-fail
+//@ only-x86_64-unknown-linux-gnu
+//@ dont-check-compiler-stderr
+//@ dont-check-compiler-stdout
+//@ ignore-backends: gcc
+
+// https://github.com/rust-lang/rust/issues/151950
+
+unsafe extern "C" {
+    #[link_name = "memset]; mov eax, 1; #"]
+    unsafe fn inject();
+}
+
+#[unsafe(export_name = "memset]; mov eax, 1; #")]
+extern "C" fn inject_() {}
+
+#[unsafe(naked)]
+extern "C" fn print_0() -> usize {
+    core::arch::naked_asm!("lea rax, [{}]", "ret", sym inject)
+}
+
+#[unsafe(naked)]
+extern "C" fn print_1() -> usize {
+    core::arch::naked_asm!("lea rax, [{}]", "ret", sym inject_)
+}
+
+fn main() {
+    dbg!(print_0());
+    dbg!(print_1());
+}
+
+//~? ERROR linking

